plans/example-browser-node/browser/index.js

```diff
@@ -0,0 +1,105 @@
+const { chromium, firefox, webkit } = require('playwright')
+const { exit } = require('process')
+const { spawn } = require('child_process')
+
+const spawnServer = require('../server')
+
+;(async () => {
+  spawnServer(8080)
+
+  let browser
+  try {
+    const browserDebugPort = process.env.BROWSER_DEBUG_PORT || 9222
+
+    switch (process.env.BROWSER_KIND || 'chromium') {
+      case 'chromium':
+        console.log(`launching chromium browser with exposed debug port: ${browserDebugPort}`)
+        browser = await chromium.launch({
+          args: [
+            '--remote-debugging-address=0.0.0.0',
+            `--remote-debugging-port=${browserDebugPort}`
+          ]
+        })
+        break
+
+      case 'firefox':
+        const localBrowserDebugPort = Number(browserDebugPort) + 1
+        console.log(`launching firefox browser with exposed debug port: ${browserDebugPort} (local ${localBrowserDebugPort})`)
+        browser = await firefox.launch({
+          args: [
+            `-start-debugger-server=${localBrowserDebugPort}`
+          ]
+        })
+
+        console.log('launching tcp proxy to expose firefox debugger for remote access')
+        const tcpProxy = spawn(
+          'socat', [
+            `tcp-listen:${browserDebugPort},bind=0.0.0.0,fork`,
+            `tcp:localhost:${localBrowserDebugPort}`
+          ]
+        )
+        tcpProxy.stdout.on('data', (data) => {
+          console.log(`firefox debugger: tcpProxy: stdout: ${data}`)
+        })
+
+        tcpProxy.stderr.on('data', (data) => {
+          console.error(`firefox debugger: tcpProxy: stderr: ${data}`)
+        })
+
+        break
+
+      case 'webkit':
+        console.log('launching webkit browser (remote debugging not yet supported)')
+        browser = await webkit.launch()
+        break
+    }
+
+    const page = await browser.newPage()
+
+    page.on('console', (message) => {
+      const loc = message.location()
+      console.log(`[${message.type()}] ${loc.url}@L${loc.lineNumber}:C${loc.columnNumber}: ${message.text()} — ${message.args()}`)
+    })
+
+    console.log('prepare page window (global) environment')
+    await page.addInitScript((nodeEnv) => {
+      const env = {}
+      for (const k in nodeEnv) {
+        if (k.startsWith('TEST_')) {
+          env[k] = nodeEnv[k]
+        }
+      }
+      window.testground = { env }
+    }, process.env)
+
+    console.log('opening up testplan webpage on localhost')
+    await page.goto('http://127.0.0.1:8080')
+
+    // TODO: wait for the test to finish somehow :|
```

Comment:
We could set a global flag in the window and use https://playwright.dev/docs/api/class-page#page-wait-for-function to detect when it's toggled. Add the toggle after `await invokeMap(testcases)` for testing, then maybe move it to the js SDK, at the end of `invokeMap`.

I wish we could rely on the SDK here and listen to the success / failure events (https://github.com/testground/sdk-go/blob/master/runtime/runenv_events.go#L13-L19), but these events do not contain the instance ID, and we might lose metrics and files written AFTER we called RecordSucess.

Reply:
Nice, will look into it. Thanks!

Reply:
Added a hook in https://github.com/testground/sdk-js/pull/26/commits/03fc268de60f1cc4286ec490fe5ef9eb9e402263, which will make sure `window.testground.result` is defined at the end. If we ever want we can distinguish between success (`true`) and something else (e.g. `Error` for a failure), but for now I just wait until it is defined: <https://github.com/testground/testground/pull/1502/commits/77e20fe4de06b1fea81f08e75a290a5caf3e54fc>